server/src/services/event/EventService.js

```diff
@@ -1,27 +1,62 @@
 const dayjs = require('dayjs');
 const timeSpans = require('./timeSpans');
 const scopes = require('./scopes');
+const AnalyzerService = require('./AnalyzerService');
+const ActionService = require('./ActionService');
 
 class EventService {
     constructor(prefix, redisService) {
         this.prefix = prefix;
         this.redisService = redisService;
+        this.actionService = new ActionService(prefix, redisService);
     }
 
     get stores() {
         const redisService = this.redisService;
+        const actionService = this.actionService;
 
         return [
             function (prefix, key, userId) {
                 return redisService.setBit(`${prefix}:bitmap:${key}`, userId, 1);
             },
 
-            async function (prefix, key) {
+            function (prefix, key) {
                 return redisService.increment(`${prefix}:increment:${key}`);
             },
 
-            async function (prefix, key, userId) {
+            function (prefix, key, userId) {
                 return redisService.addToSet(`${prefix}:set:${key}`, userId);
+            },
+
+            async function (prefix, key, userId, { action, date }) {
+                if (!action || !['register', 'buy'].includes(action) || !key.startsWith('global')) {
+                    return;
+                }
+
+                const actionCount =
+                    action === 'buy'
+                        ? await actionService.countBefore(userId, 'register', date)
+                        : await actionService.countAfter(userId, 'buy', date);
+
+                if (actionCount < 1) {
+                    return;
+                }
+
+                return redisService.setBit(`${prefix}:cohort:${key}`, userId, 1);
+            },
+
+            async function (prefix, key, userId, { action }) {
+                if (!action || action !== 'buy' || !key.startsWith('global')) {
+                    return;
+                }
+
+                const actionCount = await actionService.count(userId, 'buy');
```

Comment:
The same here,
we need just count bitmap {prefix}-bitmap-buy-{timespan today}
if 0 then
count bitmap {prefix}-bitmap-buy-anytime
if 1 then
add to retention list {prefix}-list-retention:buy-{timestamp}